csharp/src/Ice/ConnectionFactory.cs

```diff
@@ -95,43 +92,115 @@ internal sealed class OutgoingConnectionFactory
                     }
                 }
 
-                // Try to find a connection matching one of the connector.
-                foreach ((IConnector connector, Endpoint _, string _) in connectors)
+                // Wait for the first successful connection establishment
+                Task<Connection> completedTask;
+                do
                 {
-                    ValueTuple<IConnector, string> key = (connector, connectionId);
-                    if (!_pending.ContainsKey(key) &&
-                        _connectionsByConnector.TryGetValue(key, out ICollection<Connection>? connectionList))
+                    completedTask = await Task.WhenAny(connectTasks).ConfigureAwait(false);
+                    if (completedTask.IsCompletedSuccessfully)
                     {
-                        foreach (Connection connection in connectionList)
+                        lock (_mutex)
                         {
-                            if (connection.Active) // Don't return destroyed or un-validated connections
+                            Connection connection = completedTask.Result;
+                            foreach ((IConnector connector, Endpoint endpoint) in connectors)
                             {
-                                return connection;
+                                // If the connection was established for another endpoint but to the same connector,
+                                // we ensure to also associate the connection with this endpoint.
+                                if (connection.Connector.Equals(connector) && !connection.Endpoints.Contains(endpoint))
+                                {
+                                    Debug.Assert(connection.ConnectionId == connectionId);
+                                    connection.Endpoints.Add(endpoint);
+                                    _connectionsByEndpoint.Add((endpoint, connectionId), connection);
+                                    break;
+                                }
                             }
+                            return connection;
                         }
                     }
+                    connectTasks.Remove(completedTask);
                 }
-            }
+                while (connectTasks.Count > 0);
 
-            // Wait for connection establishment to one or some of the connectors to complete.
-            return await WaitForConnectOrConnectAsync(connectors, hasMore).ConfigureAwait(false);
+                // Remove the connectors we tried from the set of remaining connectors
+                connectors.RemoveAll(((IConnector Connector, Endpoint Endpoint) tuple) =>
+                    tried.Contains(tuple.Connector));
+
+                // If there are no more connectors to try, we failed to establish a connection and we raise the
+                // failure.
+                if (connectors.Count == 0)
+                {
+                    return await completedTask.ConfigureAwait(false);
```

Comment:
Can we assert here that task failed

```suggestion
                    Debug.Assert(completedTask.IsFaulted );
                    return await completedTask.ConfigureAwait(false);
```